### ISSUE TITLE
Test failure: CheckStdoutLoggingToPipeWithFirstWrite - ServerManager retries and exceptions

### ISSUE THREAD

**Issue by ryanbrandenburg** [aspnet/IISIntegration#1156]
This test [fails](http://aspnetci/viewLog.html?buildId=523284&buildTypeId=XPlat_Windows_Win2008r2_Universe) occasionally with the following error:
```
System.AggregateException : One or more errors occurred. (Filename: \\?\C:\Windows\system32\inetsrv\config\applicationHost.config
Error: Cannot write configuration file
) (Unable to load DLL 'api-ms-win-service-core-l1-1-1.dll' or one of its dependencies: The specified module could not be found. (Exception from HRESULT: 0x8007007E))
---- System.Runtime.InteropServices.COMException : Filename: \\?\C:\Windows\system32\inetsrv\config\applicationHost.config
Error: Cannot write configuration file
---- System.DllNotFoundException : Unable to load DLL 'api-ms-win-service-core-l1-1-1.dll' or one of its dependencies: The specified module could not be found. (Exception from HRESULT: 0x8007007E)
----- Inner Stack Trace #1 (System.Runtime.InteropServices.COMException) -----
   at Microsoft.Web.Administration.Interop.IAppHostWritableAdminManager.CommitChanges()
   at Microsoft.Web.Administration.ConfigurationManager.CommitChanges()
   at Microsoft.Web.Administration.ServerManager.CommitChanges()
   at Microsoft.AspNetCore.Server.IntegrationTesting.IISApplication.StartIIS(Uri uri, String contentRoot) in /_/src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISApplication.cs:line 77
   at Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer.DeployAsync() in /_/src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs:line 85
   at Microsoft.AspNetCore.Server.IntegrationTesting.FunctionalTestsBase.DeployAsync(IISDeploymentParameters parameters) in /_/test/Common.FunctionalTests/Utilities/FunctionalTestsBase.cs:line 42
   at Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests.LoggingTests.CheckStdoutLoggingToPipeWithFirstWrite(String path) in /_/test/Common.FunctionalTests/Inprocess/LoggingTests.cs:line 144
--- End of stack trace from previous location where exception was thrown ---
----- Inner Stack Trace #2 (System.DllNotFoundException) -----
   at Interop.mincore.EnumServicesStatusEx(IntPtr databaseHandle, Int32 infolevel, Int32 serviceType, Int32 serviceState, IntPtr status, Int32 size, Int32& bytesNeeded, Int32& servicesReturned, Int32& resumeHandle, String group)
   at System.ServiceProcess.ServiceController.GenerateNames()
   at System.ServiceProcess.ServiceController.GetServiceHandle(Int32 desiredAccess)
   at System.ServiceProcess.ServiceController.GenerateStatus()
   at Microsoft.Web.Administration.ServerManager.GetServiceStatus(String serviceName)
   at Microsoft.Web.Administration.ApplicationPool.Stop()
   at Microsoft.AspNetCore.Server.IntegrationTesting.IISApplication.StopAppPool() in /_/src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISApplication.cs:line 307
   at Microsoft.AspNetCore.Server.IntegrationTesting.IISApplication.StopAndDeleteAppPool() in /_/src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISApplication.cs:line 139
   at Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer.Dispose() in /_/src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs:line 36
   at Microsoft.AspNetCore.Server.IntegrationTesting.FunctionalTestsBase.StopServer() in /_/test/Common.FunctionalTests/Utilities/FunctionalTestsBase.cs:line 53
   at ReflectionAbstractionExtensions.DisposeTestClass(ITest test, Object testClass, IMessageBus messageBus, ExecutionTimer timer, CancellationTokenSource cancellationTokenSource) in C:\Dev\xunit\xunit\src\xunit.execution\Extensions\ReflectionAbstractionExtensions.cs:line 76
------- Stdout: -------
| [0.001s] TestLifetime Information: Starting test CheckStdoutLoggingToPipeWithFirstWrite-CheckLogFile at 2018-08-03T18:14:04
| [0.004s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Information: Deploying [Variation] :: ServerType=IIS, Runtime=CoreClr, Arch=x64, BaseUrlHint=, Publish=True
| [0.005s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Debug: Processing C:\BuildAgent\temp\buildTmp\80f95ea85c36440e9240e0f9d7bba463\runtimes\unix\lib\netcoreapp2.0
| [0.005s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Debug:   Copying System.Security.Principal.Windows.dll
| [0.007s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Debug: Processing C:\BuildAgent\temp\buildTmp\80f95ea85c36440e9240e0f9d7bba463\runtimes\unix\lib
| [0.007s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Debug: Processing C:\BuildAgent\temp\buildTmp\80f95ea85c36440e9240e0f9d7bba463\runtimes\unix
| [0.008s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Debug: Processing C:\BuildAgent\temp\buildTmp\80f95ea85c36440e9240e0f9d7bba463\runtimes\win\lib\netcoreapp2.0
| [0.008s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Debug:   Copying System.Security.Principal.Windows.dll
| [0.010s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Debug: Processing C:\BuildAgent\temp\buildTmp\80f95ea85c36440e9240e0f9d7bba463\runtimes\win\lib
| [0.010s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Debug: Processing C:\BuildAgent\temp\buildTmp\80f95ea85c36440e9240e0f9d7bba463\runtimes\win
| [0.010s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Debug: Processing C:\BuildAgent\temp\buildTmp\80f95ea85c36440e9240e0f9d7bba463\runtimes
| [0.010s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Debug: Processing C:\BuildAgent\temp\buildTmp\80f95ea85c36440e9240e0f9d7bba463\wwwroot
| [0.013s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Debug:   Copying static.txt
| [0.014s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Debug: Processing C:\BuildAgent\temp\buildTmp\80f95ea85c36440e9240e0f9d7bba463\x64
| [0.014s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Debug:   Copying aspnetcorev2_inprocess.dll
| [0.015s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Debug:   Copying aspnetcorev2_inprocess.pdb
| [0.023s] Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer Debug: 
...
```
Other tests within that build may have failed with a similar message, but they are not listed here. Check the link above for more info.

CC @Eilon,@muratg,@mkArtakMSFT

This issue was made automatically. If there is a problem contact @ryanbrandenburg.

**Comment by ryanbrandenburg** [comment]
Please use this workflow to address this flaky test issue, including checking applicable checkboxes and filling in the applicable "TODO" entries:

* Is this actually a flaky test?
  * No, this is a regular test failure, fix the test/product (TODO: Link to commit/PR)
  * Yes, proceed below...

* Is this test failure caused by product code flakiness? (Either this product, or another product this test depends on.)
  * [ ] File a bug against the product (TODO: Link to other bug)
  * Is it possible to change the test to avoid the flakiness?
    * Yes? Go to the "Change the test!" section.
    * No?
      * [ ] Disable the test (TODO: Link to PR/commit)
      * [ ] Wait for other bug to be resolved
      * [ ] Wait for us to get build that has the fix
      * [ ] Re-enable our test (TODO: Link to PR/commit)
      * [ ] Close this bug

* Is it that the test itself is flaky? This includes external transient problems (e.g. remote server problems, file system race condition, etc.)
  * Is there is a way to change our test to avoid this flakiness?
    * Yes? Change the test!
      * [ ] Change the test to avoid the flakiness, for example by using a different test strategy, or by adding retries w/ timeouts (TODO: Link to PR/commit)
      * [ ] Run the test 100 times locally as a sanity check.
      * [ ] Close this bug
    * No?
      * [ ] Delete the test because flaky tests are not useful (TODO: Link to PR/commit)

This comment was made automatically. If there is a problem contact @ryanbrandenburg.

**Comment by ryanbrandenburg** [comment]
CheckStdoutLoggingToPipeWithFirstWrite [failed with about the same error](http://aspnetci/viewLog.html?buildId=524530&buildTypeId=XPlat_Windows_Win8_Universe).

This comment was made automatically. If there is a problem contact @ryanbrandenburg.

**Comment by mikeharder** [comment]
For the first failure, we believe the fix is to add a retry loop around `ServerManager.CommitChanges()`:

```
---- System.Runtime.InteropServices.COMException : Filename: \\?\C:\Windows\system32\inetsrv\config\applicationHost.config
Error: Cannot write configuration file
----- Inner Stack Trace #1 (System.Runtime.InteropServices.COMException) -----
   at Microsoft.Web.Administration.Interop.IAppHostWritableAdminManager.CommitChanges()
   at Microsoft.Web.Administration.ConfigurationManager.CommitChanges()
   at Microsoft.Web.Administration.ServerManager.CommitChanges()
```

For the second failure, the fix is to catch all exceptions rather than whitelisting a few (which currently excludes `FileLoadException`):

```
---- System.IO.FileLoadException : The process cannot access the file because it is being used by another process. (Exception from HRESULT: 0x80070020)
----- Inner Stack Trace #1 (System.IO.FileLoadException) -----
   at Microsoft.Web.Administration.Interop.IAppHostMethodInstance.Execute()
   at Microsoft.Web.Administration.Site.Start()
   at Microsoft.AspNetCore.Server.IntegrationTesting.IISApplication.WaitUntilSiteStarted() in /_/src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISApplication.cs:line 109
   at Microsoft.AspNetCore.Server.IntegrationTesting.IISApplication.StartIIS(Uri uri, String contentRoot) in /_/src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISApplication.cs:line 80
   at Microsoft.AspNetCore.Server.IntegrationTesting.IIS.IISDeployer.DeployAsync() in /_/src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs:line 85
   at Microsoft.AspNetCore.Server.IntegrationTesting.FunctionalTestsBase.DeployAsync(IISDeploymentParameters parameters) in /_/test/Common.FunctionalTests/Utilities/FunctionalTestsBase.cs:line 42
   at Microsoft.AspNetCore.Server.IISIntegration.FunctionalTests.LoggingTests.CheckStdoutLoggingToPipeWithFirstWrite(String path) in /_/test/Common.FunctionalTests/Inprocess/LoggingTests.cs:line 144
--- End of stack trace from previous location where exception was thrown ---
```

https://github.com/aspnet/IISIntegration/blob/e6330ab19bdf058953b7547ef04aae7034b21658/src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS/IISDeployer.cs#L269-L274

**Comment by ryanbrandenburg** [comment]
CheckStdoutLoggingToPipe_DoesNotCrashProcess [failed with about the same error](http://aspnetci/viewLog.html?buildId=526879&buildTypeId=XPlat_Windows_Win7_Universe).

This comment was made automatically. If there is a problem contact @ryanbrandenburg.

**Comment by pakrym** [comment]
Fixed in #1197. With some follow up in https://github.com/aspnet/IISIntegration/pull/1207
